File: CS/Shared/Common.DataModel.EntityFramework/EntityFrameworkDataProvider.cs
using DevExpress.Xpf.Data;
using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Linq.Expressions;

namespace DevExpress.CRUD.DataModel.EntityFramework {
    public class EntityFrameworkDataProvider<TContext, TEntity, T> : IDataProvider<T>
        where TEntity : class
        where T : class
        where TContext : DbContext {

        protected readonly Func<TContext> createContext;
        protected readonly Func<TContext, DbSet<TEntity>> getDbSet;
        readonly Expression<Func<TEntity, T>> getEntityExpression;
        readonly string keyProperty;

        public EntityFrameworkDataProvider(Func<TContext> createContext, Func<TContext, DbSet<TEntity>> getDbSet, Expression<Func<TEntity, T>> getEnityExpression, string keyProperty) {
            this.createContext = createContext;
            this.getDbSet = getDbSet;
            this.getEntityExpression = getEnityExpression;
            this.keyProperty = keyProperty;
        }

        IList<T> IDataProvider<T>.Read() {
            using(var context = createContext()) {
                var query = getDbSet(context)
                    .Select(getEntityExpression);
                return query.ToList();
            }
        }

        IList<T> IDataProvider<T>.Fetch(SortDefinition[] sortOrder, Expression<Func<T, bool>> filter, int skip, int take) {
            using(var context = createContext()) {
                var query = getDbSet(context)
                    .Select(getEntityExpression)
                    .SortBy(sortOrder, defaultUniqueSortPropertyName: keyProperty)
                    .Where(filter)
                    .Skip(skip)
                    .Take(take);
                return query.ToList();
            }
        }
        object[] IDataProvider<T>.GetTotalSummaries(SummaryDefinition[] summaries, Expression<Func<T, bool>> filter) {
            using(var context = createContext()) {
                var query = getDbSet(context)
                    .Select(getEntityExpression)
                    .Where(filter);
                return query.GetSummaries(summaries);
            }
        }

        ValueAndCount[] IDataProvider<T>.GetDistinctValues(string propertyName, Expression<Func<T, bool>> filter) {
            using(var context = createContext()) {
                var query = getDbSet(context)
                    .Select(getEntityExpression)
                    .Where(filter);
                return query.DistinctWithCounts(propertyName);
            }
        }
    }
}

File: VB/GridControlCRUDMVVModel.EntityFramework/EntityFrameworkDataProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Data.Entity
Imports System.Linq
Imports System.Linq.Expressions

Namespace DevExpress.CRUD.DataModel.EntityFramework

    Public Class EntityFrameworkDataProvider(Of TContext As DbContext, TEntity As Class, T As Class)
        Implements IDataProvider(Of T)

        Protected ReadOnly createContext As Func(Of TContext)

        Protected ReadOnly getDbSet As Func(Of TContext, DbSet(Of TEntity))

        Private ReadOnly getEntityExpression As Expression(Of Func(Of TEntity, T))

        Public Sub New(ByVal createContext As Func(Of TContext), ByVal getDbSet As Func(Of TContext, DbSet(Of TEntity)), ByVal getEnityExpression As Expression(Of Func(Of TEntity, T)))
            Me.createContext = createContext
            Me.getDbSet = getDbSet
            getEntityExpression = getEnityExpression
        End Sub

        Private Function Read() As IList(Of T) Implements IDataProvider(Of T).Read
            Using context = createContext()
                Dim query = getDbSet(context).[Select](getEntityExpression)
                Return query.ToList()
            End Using
        End Function
    End Class
End Namespace
